### PR TITLE
Dwnl localize

### PR DESCRIPTION
**Pull request by milawynsrealm**
## Purpose

Localize the file download utility.

JIRA issue: [CORE-14498](https://jira.reactos.org/browse/CORE-14498)

## Proposed changes

- Make it so that the file download utility can be translated into other languages.
- Also, added a timestamp feature that tells the user when the download started similar to how wget works (e.g. --2018-9-12 7:43:55--). This can be reverted on request.

### CODE REVIEW

File: base/applications/network/dwnl/dwnl.c [hunk 1]
```diff
@@ -336,23 +344,24 @@ get_display_url(IN LPURL_COMPONENTS purl,
 }
 
 static int
-download_file(IN LPCTSTR pszUrl,
-              IN LPCTSTR pszFile  OPTIONAL)
+download_file(IN LPCWSTR pszUrl,
+             IN LPCWSTR pszFile OPTIONAL)
```
Comment (binarymaster) [inline base/applications/network/dwnl/dwnl.c]: One space character is missing for proper indentation.

File: base/applications/network/dwnl/dwnl.c [hunk 2]
```diff
@@ -411,7 +420,9 @@ download_file(IN LPCTSTR pszUrl,
     if (iRet <= 0)
         return iRet;
 
-    _tprintf(_T("Download `%s\'\n\t=> `%s\'\n"), szUrl, pszFile);
+    GetLocalTime(&sysTime);
+
+    ConPrintf(StdOut, "--%d-%d-%d %d:%d:%d-- ""%s""\n\t=> ""%s""\n", sysTime.wYear, sysTime.wMonth, sysTime.wDay, sysTime.wHour, sysTime.wMinute, sysTime.wSecond, szUrl, pszFile);
```
Comment (binarymaster) [inline base/applications/network/dwnl/dwnl.c]:
```
../../base/applications/network/dwnl/dwnl.c: In function 'download_file':
../../base/applications/network/dwnl/dwnl.c:425:5: error: passing argument 2 of 'ConPrintf' from incompatible pointer type [-Werror]
```

Comment (milawynsrealm) [reply base/applications/network/dwnl/dwnl.c]: Fixed with latest commit.

File: base/applications/network/dwnl/dwnl.c (hunk 1 above)
Comment (milawynsrealm) [reply base/applications/network/dwnl/dwnl.c]: Fixed.

File: base/applications/network/dwnl/dwnl.c [hunk 3]
```diff
@@ -336,23 +344,24 @@ get_display_url(IN LPURL_COMPONENTS purl,
 }
 
 static int
-download_file(IN LPCTSTR pszUrl,
-              IN LPCTSTR pszFile  OPTIONAL)
+download_file(IN LPCWSTR pszUrl,
+             IN LPCWSTR  pszFile OPTIONAL)
```
Comment (binarymaster) [inline base/applications/network/dwnl/dwnl.c]: Space should be before `IN`, and fix double space before `pszFile`.

**Comment by milawynsrealm** [comment]
Now that all the checks have finally passed, how would I squash all the commits into one?

**Comment by binarymaster** [comment]
> @milawynsrealm wrote:
> how would I squash all the commits into one?

Here are git commands for your current PR:
```
git reset --hard 6b67ef6e101ab1b1d98d3c285213cde96628deda
git cherry-pick -n 8e88740727e5b3679272f160c69bcd4ef8cbd557
git cherry-pick -n bc837e9197e1dec1d09fbbc69d942589b3aa57a8
git cherry-pick -n 56a6c77e58284e48d6955341a815a824a4d70c1d
git cherry-pick -n e955b084dca425a467be6ed150bc596f86852a3e
git cherry-pick -n 335c8d3a4e54b64c0ce3c70d87ea22f26bce7938
git commit -a --author "Lee Schroeder <spaceseel@gmail.com>" --date="Tue Sep 11 17:59:10 2018 -0600"
git push --force-with-lease
```
Editor will appear after `git commit`, enter proper commit message into it.

**Comment by milawynsrealm** [comment]
C:\Games\MyCode\reactos>git reset --hard 6b67ef6e101ab1b1d98d3c285213cde96628deda
fatal: Could not parse object '6b67ef6e101ab1b1d98d3c285213cde96628deda'.

**Comment by binarymaster** [comment]
Ok, then you should also do:
```
git fetch upstream
```
or
```
git fetch reactos
```

Depends on how your upstream remote is named.

File: base/applications/network/dwnl/dwnl.c [hunk 4]
```diff
@@ -51,17 +54,19 @@ write_status(LPCTSTR lpFmt, ...)
 
     if (GetConsoleScreenBufferInfo(GetStdHandle(STD_OUTPUT_HANDLE), &csbi))
```
Comment (HBelusca) [inline base/applications/network/dwnl/dwnl.c]: Note to /self: similar stuff is already dealt by conutils.

File: base/applications/network/dwnl/dwnl.c [hunk 5]
```diff
@@ -71,12 +76,16 @@ CBindStatusCallback_UpdateProgress(CBindStatusCallback *This)
         if (Percentage > 99)
             Percentage = 99;
 
-        write_status(_T("%2d%% (%I64u bytes downloaded)"), Percentage, This->Progress);
+        LoadStringW(NULL, IDS_BYTES_DOWNLOADED_FULL, szMessage, MAX_PATH);
```
Comment (HBelusca) [inline base/applications/network/dwnl/dwnl.c]: Replace MAX_PATH by either `ARRAYSIZE(szMessage)` or `_countof(szMessage)`, whichever one compiles successfully.

File: base/applications/network/dwnl/dwnl.c [hunk 6]
```diff
@@ -71,12 +76,16 @@ CBindStatusCallback_UpdateProgress(CBindStatusCallback *This)
...
+
+        write_status(szMessage, Percentage, This->Progress);
     }
     else
     {
+        LoadStringW(NULL, IDS_BYTES_DOWNLOADED, szMessage, MAX_PATH);
```
Comment (HBelusca) [inline base/applications/network/dwnl/dwnl.c]: Ditto.

File: base/applications/network/dwnl/dwnl.c [hunk 7]
```diff
@@ -168,75 +177,74 @@ CBindStatusCallback_OnProgress(IBindStatusCallback *iface,
...
             if (This->bResolving)
             {
-                _tprintf(_T("done.\n"));
-                _tprintf(_T("Connecting to %s[%s]... "), This->szHostName, szStatusText);
+                ConResPrintf(StdOut, IDS_DONE);
+                ConResPrintf(StdOut, IDS_CONNECTING_TO_FULL, This->szHostName, szStatusText);
             }
             else
-                _tprintf(_T("Connecting to %s... "), szStatusText);
+                ConResPrintf(StdOut, IDS_CONNECTING_TO, szStatusText);
```
Comment (HBelusca) [inline base/applications/network/dwnl/dwnl.c]: Please put this 'else' block inside curly brackets.

File: base/applications/network/dwnl/dwnl.c [hunk 8]
```diff
@@ -168,75 +177,74 @@ CBindStatusCallback_OnProgress(IBindStatusCallback *iface,
...
-                    _tprintf(_T("Length: %I64u [%s]\n"), This->Size, This->szMimeType);
+                if (This->szMimeType[0] != L'\0')
+                    ConResPrintf(StdOut, IDS_LENGTH_FULL, This->Size, This->szMimeType);
                 else
-                    _tprintf(_T("Length: %I64u\n"), This->Size);
+                    ConResPrintf(StdOut, IDS_LENGTH, This->Size);
             }
 
-            _tprintf(_T("\n"));
+            wprintf(L"\n");
```
Comment (HBelusca) [inline base/applications/network/dwnl/dwnl.c]: `ConPuts(StdOut, L"\n");`

File: base/applications/network/dwnl/dwnl.c [hunk 9]
```diff
@@ -411,7 +420,9 @@ download_file(IN LPCTSTR pszUrl,
     if (iRet <= 0)
         return iRet;
 
-    _tprintf(_T("Download `%s\'\n\t=> `%s\'\n"), szUrl, pszFile);
+    GetLocalTime(&sysTime);
+
+    ConPrintf(StdOut, L"--%d-%d-%d %d:%d:%d-- %s\n\t=> %s\n", sysTime.wYear, sysTime.wMonth, sysTime.wDay, sysTime.wHour, sysTime.wMinute, sysTime.wSecond, szUrl, pszFile);
```
Comment (HBelusca) [inline base/applications/network/dwnl/dwnl.c]:
```c
ConPrintf(StdOut, L"--%d-%d-%d %d:%d:%d-- %s\n\t=> %s\n",
          sysTime.wYear, sysTime.wMonth, sysTime.wDay,
          sysTime.wHour, sysTime.wMinute, sysTime.wSecond,
          szUrl, pszFile);

```

**Comment by milawynsrealm** [comment]
Made the changes based on HBelusca's requests. Also compiled it to makes sure that the changes worked.

File: base/applications/network/dwnl/dwnl.c (hunk 4 above)
Comment (milawynsrealm) [reply base/applications/network/dwnl/dwnl.c]: Is this something that I'm supposed to do anything with?

Comment (HBelusca) [reply base/applications/network/dwnl/dwnl.c]: Don't worry, keep the code as it is now, I'll adapt it later if really needed.

File: base/applications/network/dwnl/dwnl.c (hunk 5 above)
Comment (milawynsrealm) [reply base/applications/network/dwnl/dwnl.c]: Done

**Comment by binarymaster** [comment]
@milawynsrealm I've updated the command list to squash everything into one commit: https://github.com/reactos/reactos/pull/853#issuecomment-420994972

**Comment by milawynsrealm** [comment]
Still having trouble with the GIT commands. Can't do anything else today since I have a lot to do. Will get to it when I can.

**Comment by milawynsrealm** [comment]
> C:\Games\MyCode\reactos>git fetch reactos
fatal: 'reactos' does not appear to be a git repository
fatal: Could not read from remote repository.

Please make sure you have the correct access rights
and the repository exists.

C:\Games\MyCode\reactos>git checkout dwnl_localize
Already on 'dwnl_localize'
M       base/setup/lib/bldrsup.c
M       base/setup/lib/filesup.c
M       base/setup/lib/filesup.h
M       base/setup/usetup/bootsup.c
M       base/setup/usetup/usetup.c
M       dll/shellext/netshell/res/neterror.ico
M       dll/shellext/netshell/res/netidle.ico
M       dll/shellext/netshell/res/netoff.ico
M       dll/shellext/netshell/res/netrec.ico
M       dll/shellext/netshell/res/nettrans.ico
M       dll/shellext/netshell/res/nettrrec.ico
Your branch is up to date with 'origin/dwnl_localize'.
> C:\Games\MyCode\reactos>git reset --hard 6b67ef6e101ab1b1d98d3c285213cde96628deda
fatal: Could not parse object '6b67ef6e101ab1b1d98d3c285213cde96628deda'.

C:\Games\MyCode\reactos>git cherry-pick -n 8e88740727e5b3679272f160c69bcd4ef8cbd557
error: could not apply 8e88740727... [dwnl] Localize the File Download Utility. Also, changed IDS_DOWNLOAD to include a start time stamp similar to the one in wget.
hint: after resolving the conflicts, mark the corrected paths
hint: with 'git add <paths>' or 'git rm <paths>'

C:\Games\MyCode\reactos>git cherry-pick -n bc837e9197e1dec1d09fbbc69d942589b3aa57a8
base/applications/network/dwnl/dwnl.c: unmerged (827565367e89914fb055816b399204cee25f0973)
base/applications/network/dwnl/dwnl.c: unmerged (4de7eda54e2d41f4cfe0e302fab10c97b2538e52)
base/applications/network/dwnl/dwnl.c: unmerged (0176a976a964695e49de5a446e5530db65c21a11)
error: your index file is unmerged.
fatal: cherry-pick failed

Still having problems. Either I'm doing something wrong, or the tools are giving me problems. Guessing, It's probably the former.

**Comment by HBelusca** [comment]
What does `git remote -v` show?

**Comment by binarymaster** [comment]
> Either I'm doing something wrong, or the tools are giving me problems.

1. Don't do cherry-pick if at least one previous action is failed, this may lead to messing everything up.
2. Looks like you have uncommited/unstaged changes (setup, shellext), `git reset --hard` will completely remove them. To avoid that, use `git stash` just before reset.

**Comment by milawynsrealm** [comment]
> What does git remote -v show?

>C:\Games\MyCode\reactos>git remote -v
origin  https://github.com/milawynsrealm/reactos.git (fetch)
origin  https://github.com/milawynsrealm/reactos.git (push)

**Comment by binarymaster** [comment]
So, you have no remote upstream, add it:

`git remote add reactos https://github.com/reactos/reactos`

**Comment by milawynsrealm** [comment]
Done.

**Comment by binarymaster** [comment]
Congrats @milawynsrealm, you've successfully squashed commits! :smirk: 

But your commit message still needs to be fixed:

image

**Comment by milawynsrealm** [comment]
how do I edit the message for the commit?

**Comment by binarymaster** [comment]
```
git commit --amend
// edit your commit message
git push --force-with-lease
```

**Comment by binarymaster** [comment]
@HBelusca this PR is ready to merge :relaxed:

**Pull request merged**